Feat: jet 备份脚本

YkeLit opened the pull request:
- 使用 velero 对 postgresql 进行持久卷级别备份
- 压缩打包 minio 数据并传输到远端服务器
- 查看 postgresql 备份信息，检查最近一次备份是否成功
- 设定过期时间自动清理过期备份
- 添加备份脚本使用说明
-  添加备份脚本初始化引导

fahchen commented:
commit message 不用括号
`[feat]: jet 备份脚本` -> `feat: jet 备份脚本`

fahchen commented on utils/backup/backup.sh:
```diff
@@ -0,0 +1,78 @@
+}
+
+# 备份 postgresql 数据
+PostgresqlDump () {
```
这些 function 的名字为啥是 pascal 的呢？

fahchen commented on utils/backup/backup.sh:
```diff
@@ -0,0 +1,78 @@
+  echo `date +%Y%m%d`
+}
+
+# 备份 postgresql 数据
+PostgresqlDump () {
+  TTL=$(($EXPIR_EDAY*24))
+  velero create schedule postgres \
+    --schedule "0 2 * * *" \
+    --ttl $TTL \
+    --volume-snapshot-locations volume-snapshot \
+    --storage-location backup \
+    --include-namespaces $POSTGRESQL_NAMESPACE \
+    --selector app.kubernetes.io/instance=$POSTGRESQL_INSTANCE \
+    --snapshot-volumes \
+    --default-volumes-to-fs-backup
+}
+
+# 备份 minio 数据
+MinioBackup () {
+  [ ! -d "$MINIO_BACKUP_DIR" ] && mkdir -p $MINIO_BACKUP_DIR || echo "已存在$MINIO_BACKUP_DIR"
+  BackupFileName=$(GetDateStr)-minio.tar.gz
+  tar cf $MINIO_BACKUP_DIR/$BackupFileName $MINIO_DATA_DIR
+  if [ "$REMOTE_SERVER_IP" != '' ]; then
+    echo "将新备份文件跨服务器备份到 $REMOTE_SERVER_IP ****"
+    scp -P$REMOTE_SERVER_PORT $MINIO_BACKUP_DIR/$_BackupFileName $REMOTE_SERVER_IP:$REMOTE_BACKUP_DIR/minio/
+  fi
+}
```
这个有没可能集成进 jet chart 里面，根据 values 自动执行

YkeLit replied:
minio 的备份暂时没有想到方法，postgres 的可以

fahchen commented on utils/backup/backup.sh:
```diff
@@ -0,0 +1,78 @@
+
+set -ex
+
+# postgresql 所在 namespace
+POSTGRESQL_NAMESPACE=postgresql
+
+# postgresql 示例名称
+POSTGRESQL_INSTANCE=postgresql
+
+# minio 数据与备份位置
+MINIO_DATA_DIR=/opt/apps/data/
+MINIO_BACKUP_DIR=/opt/backup/minio
+
+# 复制到备份服务器IP，为空则
+REMOTE_SERVER_IP=10.0.0.1
+REMOTE_SERVER_PORT=22
+REMOTE_BACKUP_DIR=/opt/remote-backup
+
+# 保留备份期限，单位：天
+EXPIR_EDAY=30
```
这些值都不能写死吧？

fahchen reviewed (changes requested):
这部分最后是集成进 jet 里面，不要单独在外面手动去管理，手动的操作可以提供一些查询接口，恢复接口就好了

fahchen commented on chart/values.yaml:
```diff
@@ -102,3 +104,24 @@ minioTLSSecret:
+  - production
+selector:
+  app.kubernetes.io/instance: postgresql
+storageLocation: ""
```
这个值一般填啥呀？

YkeLit replied:
这里值是根据 velero 来的

fahchen replied:
这个值是不是必须得填写？

YkeLit replied:
这个值不填写也可以，velero 在部署的时候会指定默认的storageLocation

fahchen commented on chart/templates/cronjob/minio-backup-cronjob.yaml:
```diff
@@ -0,0 +1,46 @@
+            - -c
+            args:
+            - |
+              echo "$SSH_KEY" | base64 -d > private.key && chmod 600 private.key
+              [ ! -d {{ .Values.backupDir }} || mkdir -p {{ .Values.backupDir }}
+              tar cf `date +%Y%m%d%H%M`-minio.tar.gz /data/{{ .Values.dataDir }} \
+                && scp -P {{ .Values.remotePort }} -i private.key -o "StrictHostKeyChecking=no" \
+                `date +%Y%m%d%H%M`-minio.tar.gz {{ .Values.remoteUser }}@{{ .Values.remoteIP }}:~/{{ .Values.backupDir }}
+              ssh -p {{ .Values.remotePort }} -i private.key -o "StrictHostKeyChecking=no" \
+                {{ .Values.remoteUser }}@{{ .Values.remoteIP }} \
+                "find ~/{{ .Values.backupDir }} -maxdepth 1 -type f -mtime +{{ .Values.expireDay }} -name '*.tar.gz' -exec  rm -rf {} \;"
```
这里用 minio 的 client 来操作呢，就不用登陆到某个地方去了
https://min.io/docs/minio/linux/reference/minio-mc.html

fahchen replied:
改成 mc 之后，可以再考虑 velero

YkeLit replied:
这里我想麻烦了，我其实可以挂载两个服务器的持久卷就行了

fahchen commented on chart/values.yaml:
```diff
@@ -102,3 +104,20 @@ minioTLSSecret:
 minioServicePort: 9000
+
+
+# [Velero]
+veleroSchedule: "0 1 * * *"
+includeNamespaces:
+  - production
+selector:
+  app.kubernetes.io/instance: postgresql
+storageLocation: ""
+ttl: "720h0m0s"
```
这部分变量都加 database 的前缀，或者做成 sub-keys

fahchen commented on chart/values.yaml:
```diff
@@ -102,3 +104,20 @@ minioTLSSecret:
+minioSchedule: "0 3 * * *"
+minioDataPVC: "minio-data"
+minioBackupPVC: "minio-backup"
+dataDir: "*"
+expireDay: "30"
```
这两个变量也一样，做成前缀或者 sub-keys

fahchen reviewed (changes requested):
backup 放到一个前缀或者 sub-keys，不要污染其他的

YkeLit commented:
> backup 放到一个前缀或者 sub-keys，不要污染其他的

我把他们都放到 backup 字段下，这样应该可以吧。

fahchen commented:
> > backup 放到一个前缀或者 sub-keys，不要污染其他的
> 
> 
> 
> 我把他们都放到 backup 字段下，这样应该可以吧。

可以的，改成设置了 backup 的才启动那些 job

fahchen commented on chart/templates/velero/postgresql-backup-schedule.yaml:
```diff
@@ -0,0 +1,22 @@
+{{- with .Values.backup }}
```
我咋觉得这个不能判断是否开启 backup 呢？

YkeLit replied:
当backup 控制的时候，这两个备份不会被创建

fahchen replied:
控制是什么意思？

YkeLit replied:
空值

fahchen commented on chart/values.yaml:
```diff
@@ -102,3 +104,20 @@ minioTLSSecret:
+
+# [Backup]
+backup:
+  # [Velero]
+  veleroSchedule: "0 1 * * *"
+  includeNamespaces:
+    - production
+  selector:
+    app.kubernetes.io/instance: postgresql
+  storageLocation: ""
+  ttl: "720h0m0s"
+  # [Minio Backup]
+  minioSchedule: "0 3 * * *"
+  minioDataPVC: "minio-data"
+  minioBackupPVC: "minio-backup"
+  minioDataDir: "*"
+  minioExpireDay: "30"
```
不要默认值，默认值应该不能正常工作的，需要设置正确才能工作，所以不要给默认值

YkeLit replied:
好，这个备份就变成个可选项嘛，需要的时候才加上这个功能

fahchen replied:
是的

fahchen reviewed (changes requested):
commit message 格式有些问题，重新整理下 commit。其他没问题了

https://github.com/Byzanteam/jet-deployment/pull/18#issuecomment-1628084648

YkeLit committed:
feat: jet 备份脚本

使用 velero 对 postgresql 进行持久卷级别备份
压缩打包 minio 数据并传输到远端服务器
查看 postgresql 备份信息，以及检查最近一次备份是否成功
设定过期时间自动清理过期备份

YkeLit committed:
refactor: 修改 velero 备份创建方式

添加 values 文件中 velero 相关参数
添加 schedule 编排文件
删除脚本中创建 schedule

YkeLit committed:
refactor: 使用 k8s 定时任务备份 minio 数据

创建备份 minio 定时备份
备份方式通过挂载不同主机持久卷实现
自动清理过期备份

YkeLit committed:
refactor: 调整 values 文件字段使用 sub-key

备份字段合并到 backup 字段下
修改调用的 key 名称

YkeLit committed:
docs: 添加备份相关字段说明

YkeLit committed:
feat: 添加对 `backup` 字段判断
